sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -255,7 +255,13 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
     case OverwritePartitionsDynamic(r: DataSourceV2Relation, query, _, _, Some(write)) =>
       OverwritePartitionsDynamicExec(planLater(query), refreshCache(r), write) :: Nil
 
-    case DeleteFromTable(relation, condition) =>
+    case DeleteFromTableWithFilters(r: DataSourceV2Relation, filters) =>
+      DeleteFromTableExec(r.table.asDeletable, filters.toArray, refreshCache(r)) :: Nil
+
+    case DeleteFromTable(_, _, Some(rewritePlan)) =>
+      planLater(rewritePlan) :: Nil
+
+    case DeleteFromTable(relation, condition, None) =>
```

Comment:
Is this the same as `DeleteFromTableWithFilters`? Is it possible to combine these two?

Reply:
That's a good question. Technically, `OptimizeMetadataOnlyDeleteFromTable` has almost identical logic but applies only to plans that reference `ReplaceData` and does not throw any exceptions. Another substantial difference is the new rule does the check in the optimizer while the current check is done during physical planning. 

I am looking for feedback on this one.

Reply:
Thanks for the explanation. Is it possible to change the rule `OptimizeMetadataOnlyDeleteFromTable` to make it also apply to `DeleteFromTable(relation, condition, None)`, and move the current check to there, and then use `DeleteFromTableWithFilters`?

Reply:
Yeah, I think that will be cleaner. Is everybody okay with moving the check from the physical planning to the optimizer?

Reply:
Yes. I think I would prefer just setting the row-level plan to `None` in the optimizer. Is that the plan? (If so, then you could do that in this commit rather than following up with the `DeleteFromTableWithFilters` optimization)

Reply:
When I tried this, it actually uncovered all the differences. This rule has very specific purpose and covers only a subset of the logic in `DataSourceV2Strategy`. After a closer look, I am not sure throwing exceptions in the optimizer rule is a great idea. Since most steps are done through common utils (split predicates, normalization, filter conversion), I'd consider keeping this rule simple. It is main purpose is to remove the rewrite plan when a delete can be handled using filters.

Reply:
Please, take another look and let me know what you think, @huaxingao @rdblue @cloud-fan @viirya.